Passing File arguments in an expression placeholder results in toil-wdl-runner failing to import those files

I previously thought this was something with how Toil converts strings into paths, but it looks like that functionality is good and that the cause of this error is something else.

When calling a WDL function that takes a File as an argument, putting the function call in an expression placeholder will result in an AssertionError that the file is not a local file. It looks like it happens on any function that takes a File as an argument, and passing a String normally works.

This occurs both when the file is defined inside the workflow and when the file is defined in the JSON input.

```wdl
workflow wf {
  input {}
  call t {
  }
}

task t {
  input {}
  File f = "/home/heaucques/Documents/toil-examples/test.wdl" # any file, error also happens if f is defined from the json
  command {
    echo ${size(f)} > output.txt
  }
}
```
```
[2023-06-28T15:25:08-0700] [MainThread] [E] [toil.wdl.wdltoil] Expression evaluation failed for command: 
	    echo ~{size(f)} > output.txt
	  
	Traceback (most recent call last):
	  File "/home/heaucques/Documents/toil/venv/lib/python3.10/site-packages/WDL/Expr.py", line 123, in eval
	    ans = self._eval(env, stdlib)
	  File "/home/heaucques/Documents/toil/venv/lib/python3.10/site-packages/WDL/Expr.py", line 1140, in _eval
	    return f(self, env, stdlib)
	  File "/home/heaucques/Documents/toil/venv/lib/python3.10/site-packages/WDL/StdLib.py", line 232, in __call__
	    return self._call_eager(expr, [arg.eval(env, stdlib=stdlib) for arg in expr.arguments])
	  File "/home/heaucques/Documents/toil/src/toil/wdl/wdltoil.py", line 366, in _call_eager
	    total_size += os.path.getsize(self.stdlib._devirtualize_filename(uri))
	  File "/home/heaucques/Documents/toil/src/toil/wdl/wdltoil.py", line 452, in _devirtualize_filename
	    assert os.path.exists(result), f"Virtualized file {filename} looks like a local file but isn't!"
	AssertionError: Virtualized file /mnt/miniwdl_task_container/work/_miniwdl_inputs/0/test.wdl looks like a local file but isn't!
```

┆Issue is synchronized with this [Jira Story](https://ucsc-cgl.atlassian.net/browse/TOIL-1359)
┆Issue Number: TOIL-1359

--- a/src/toil/wdl/wdltoil.py
+++ b/src/toil/wdl/wdltoil.py
@@ -523,6 +523,10 @@ class ToilWDLStdLibTaskCommand(ToilWDLStdLibBase):
         # use the out-of-container equivalent.
         result = self.container.host_path(filename)
 
+        if result is None:
+            # We really shouldn't have files in here that we didn't virtualize.
+            raise RuntimeError(f"File {filename} in container is not mounted from the host and can't be opened form the host")
+
         logger.debug('Devirtualized %s as out-of-container file %s', filename, result)
         return result
 
@@ -534,8 +538,8 @@ class ToilWDLStdLibTaskCommand(ToilWDLStdLibBase):
         """
 
         if filename not in self.container.input_path_map:
-            # Mount the file
-            task_container.add_paths([filename])
+            # Mount the file.
+            self.container.add_paths([filename])
 
         result = self.container.input_path_map[filename]
         

--- a/src/toil/wdl/wdltoil.py
+++ b/src/toil/wdl/wdltoil.py
@@ -377,6 +377,32 @@ class NonDownloadingSize(WDL.StdLib._Size):
         # Return the result as a WDL float value
         return WDL.Value.Float(total_size)
 
+
+# Both the WDL code itself **and** the commands that it runs will deal in
+# "virtualized" filenames.
+
+# We have to guarantee that "When a WDL author uses a File input in their
+# Command Section, the fully qualified, localized path to the file is
+# substituted when that declaration is referenced in the command template."
+
+# This has to be true even if the File is the result of a WDL function that is
+# run *during* the evaluation of the command string, via a placeholder
+# expression evaluation.
+
+# Really there are 3 filename spaces in play: Toil filestore URLs,
+# outside-the-container host filenames, and inside-the-container filenames. But
+# the MiniWDL machinery only gives us 2 levels to work with: "virtualized"
+# (visible to the workflow) and "devirtualized" (openable by this process).
+
+# So we sneakily swap out what "virtualized" means. Usually (as provided by
+# ToilWDLStdLibBase) a "virtualized" filename is the Toil filestore URL space.
+# But when evaluating a task command, we switch things so that the
+# "virtualized" space is the inside-the-container filename space (by
+# devirtualizing and then host-to-container-mapping all the visible files, and
+# then using ToilWDLStdLibTaskCommand for evaluating expressions, and then
+# going back from container to host space after the command). At all times the
+# "devirtualized" space is outside-the-container host filenames.
+
 class ToilWDLStdLibBase(WDL.StdLib.Base):
     """
     Standard library implementation for WDL as run on Toil.
@@ -398,14 +424,7 @@ class ToilWDLStdLibBase(WDL.StdLib.Base):
         self.size = NonDownloadingSize(self)
 
         # Keep the file store around so we can access files.
-        self._file_store = file_store
-
-    # Both the WDL code itself **and** the commands that it runs will deal in
-    # "virtualized" filenames by default, so when making commands we need to
-    # make sure to devirtualize filenames.
-    # We have to guarantee that "When a WDL author uses a File input in their
-    # Command Section, the fully qualified, localized path to the file is
-    # substituted when that declaration is referenced in the command template."
+        self._file_store = file_store 
 
     def _is_url(self, filename: str, schemes: List[str] = ['http:', 'https:', 's3:', 'gs:', TOIL_URI_SCHEME]) -> bool:
         """

--- a/src/toil/wdl/wdltoil.py
+++ b/src/toil/wdl/wdltoil.py
@@ -525,7 +525,7 @@ class ToilWDLStdLibTaskCommand(ToilWDLStdLibBase):
 
         if result is None:
             # We really shouldn't have files in here that we didn't virtualize.
-            raise RuntimeError(f"File {filename} in container is not mounted from the host and can't be opened form the host")
+            raise RuntimeError(f"File {filename} in container is not mounted from the host and can't be opened from the host")
 
         logger.debug('Devirtualized %s as out-of-container file %s', filename, result)
         return result

--- a/src/toil/wdl/wdltoil.py
+++ b/src/toil/wdl/wdltoil.py
@@ -470,6 +470,59 @@ class ToilWDLStdLibBase(WDL.StdLib.Base):
         logger.debug('Virtualized %s as WDL file %s', filename, result)
         return result
 
+class ToilWDLStdLibTaskCommand(ToilWDLStdLibBase):
+    """
+    Standard library implementation to use inside a WDL task command evaluation.
+
+    Expects all the filenames in variable bindings to be container-side paths;
+    these are the "virtualized" filenames, while the "devirtualized" filenames
+    are host-side paths.
+    """
+
+    def __init__(self, file_store: AbstractFileStore, container: TaskContainer):
+        """
+        Set up the standard library for the task command section.
+        """
+
+        # TODO: Don't we want to make sure we don't actually use the file store?
+        super().__init__(file_store)
+        self.container = container
+
+    def _devirtualize_filename(self, filename: str) -> str:
+        """
+        Go from a virtualized WDL-side filename to a local disk filename.
+
+        Any WDL-side filenames which are paths will be paths in the container. 
+        """
+        if self._is_url(filename):
+            # We shouldn't have to deal with URLs here; we want to have exactly
+            # two nicely stacked/back-to-back layers of virtualization, joined
+            # on the out-of-container paths.
+            raise RuntimeError(f"File {filename} is a URL but should already be an in-container-virtualized filename")
+        
+        # If this is a local path it will be in the container. Make sure we
+        # use the out-of-container equivalent.
+        result = self.container.host_path(filename)
+
+        logger.debug('Devirtualized %s as out-of-container file %s', filename, result)
+        return result
+
+
+    def _virtualize_filename(self, filename: str) -> str:
+        """
+        From a local path in write_dir, 'virtualize' into the filename as it should present in a
+        File value, when substituted into a command in the container.
+        """
+
+        if filename not in self.container.input_path_map:
+            # Mount the file
+            task_container.add_paths([filename])
+
+        result = self.container.input_path_map[filename]
+        
+        logger.debug('Virtualized %s as WDL file %s', filename, result)
+        return result
+
 class ToilWDLStdLibTaskOutputs(ToilWDLStdLibBase, WDL.StdLib.TaskOutputs):
     """
     Standard library implementation for WDL as run on Toil, with additional
@@ -1216,8 +1269,11 @@ class WDLTaskJob(WDLBaseJob):
             # TODO: MiniWDL deals with directory paths specially here.
             contained_bindings = map_over_files_in_bindings(bindings, lambda path: task_container.input_path_map[path])
 
+            # Make a new standard library for evaluating the command specifically, which only deals with in-container paths and out-of-container paths.
+            command_library = ToilWDLStdLibTaskCommand(file_store, task_container)
+
             # Work out the command string, and unwrap it
-            command_string: str = evaluate_named_expression(self._task, "command", WDL.Type.String(), self._task.command, contained_bindings, standard_library).coerce(WDL.Type.String()).value
+            command_string: str = evaluate_named_expression(self._task, "command", WDL.Type.String(), self._task.command, contained_bindings, command_library).coerce(WDL.Type.String()).value
 
             # Grab the standard out and error paths. MyPy complains if we call
             # them because in the current MiniWDL version they are untyped.

